### ISSUE TITLE
using gru or lstm (cuDNN implementation)

### ISSUE THREAD

**Issue by kaczmarj** [jhajagos/LSTMSeq2ML#1]
@jhajagos - i started updating the model implementation. in a code comment, you reference `CuDNNLSTM`, but in the model architecture, you used `GRU` layers. which should i use? gru or lstm? i began an updated implementation in the branch enh/tf2-cudnn:

https://github.com/jhajagos/LSTMSeq2ML/blob/0bfd37f6463a54c69f01b814bd948a318c35363d/train_model_example.py#L28-L61

fyi `CuDNNLSTM` has been deprecated in tensorflow 2.x in favor of LSTM. similar for GRU. based on the keyword arguments passed to LSTM or GRU layers (and the availability of a gpu), the cuDNN will be used transparently. i have written out the cuDNN-compatible keyword arguments in the code i linked above. the missing piece is removing masking, because for cuDNN to be used, inputs cannot be masked.

**Comment by jhajagos** [comment]
It looks like you have to remove the masking and put shape=(None, 255) for 255 variables. See Chollet's answer here:

https://groups.google.com/a/tensorflow.org/forum/#!topic/discuss/QB1Ilze5ZK0

We would probable pass in variable lengths sequences. So do we need to generate variable lengths matrices with each batch which might require using Sequence in utilities.

https://www.tensorflow.org/api_docs/python/tf/keras/utils/Sequence

We can focus on GRU for now. 

```python
def gru_model():
    m = Sequential()
    ..
   return m

model_dictionary = {"grud_model": gru_model()}

I want to create a plugin where the user can select different model on the command line and train the model.

We can meet later in the week after your test to discuss.
```

**Comment by kaczmarj** [comment]
OK, I included `input_shape=(None, 255)`. I'm not sure if this is an obvious NLP technical point, but I learned that the sequence lengths must be the same within a batch. Sequence lengths can be different between batches. To achieve this behavior, we could use `tf.keras.utils.Sequence`, but we can also use [`tf.RaggedTensor`](https://www.tensorflow.org/api_docs/python/tf/RaggedTensor) or [`tf.data.Dataset.from_generator`](https://www.tensorflow.org/api_docs/python/tf/data/Dataset#from_generator).
